### ISSUE TITLE
GpuArrayException: ('cuLinkAddData: CUDA_ERROR_UNKNOWN: unknown error', 3)

### ISSUE THREAD

**Issue by chcgnzls** [Theano/Theano#6613]
I am trying to run theano 1.0.2 on a gpu cluster and am also using CUDA 8.0, pygpu 0.7.6, and python 2.7 on CentOS 6.8

I get the following error when importing theano with the flags `THEANO_FLAGS='device=cuda,dnn.enabled=False'`:

```python
 >>> import theano
Can not use cuDNN on context None: Disabled by dnn.enabled flag
ERROR (theano.gpuarray): Could not initialize pygpu, support disabled
Traceback (most recent call last):
  File "/home/sergiog/.conda/envs/dm/lib/python2.7/site-packages/theano/gpuarray/__init__.py", line 227, in <module>
    use(config.device)
  File "/home/sergiog/.conda/envs/dm/lib/python2.7/site-packages/theano/gpuarray/__init__.py", line 214, in use
    init_dev(device, preallocate=preallocate)
  File "/home/sergiog/.conda/envs/dm/lib/python2.7/site-packages/theano/gpuarray/__init__.py", line 159, in init_dev
    pygpu.blas.gemm(0, tmp, tmp, 0, tmp, overwrite_c=True)
  File "pygpu/blas.pyx", line 149, in pygpu.blas.gemm
  File "pygpu/blas.pyx", line 47, in pygpu.blas.pygpu_blas_rgemm
GpuArrayException: ('cuLinkAddData: CUDA_ERROR_UNKNOWN: unknown error', 3)
```
Output of `nvidia-smi`:

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 387.26                 Driver Version: 387.26                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla K20Xm         Off  | 00000000:02:00.0 Off |                    0 |
| N/A   25C    P0    62W / 235W |      0MiB /  5699MiB |    100%   E. Process |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```

The issue seems very similar to [#6531](https://github.com/Theano/Theano/issues/653) but I do not have two versions of theano installed in my environment after checking `conda list` and rebuilding several times.

**Comment by abergeron** [comment]
This is not an error I have encountered often.  The only hints I've gathered point to a partly broken CUDA install.  How did you install it?

**Comment by chcgnzls** [comment]
Is there anything in particular that make you think it's a broken installation?
CUDA was installed by the cluster system admin so I can't troubleshoot/debug that directly.

**Comment by abergeron** [comment]
Just the error message.  I've seen that from a couple a people before and reinstalling CUDA usually fixed it like this issue: https://github.com/Theano/libgpuarray/issues/552

**Comment by abergeron** [comment]
Apparently this can also be due to the driver version: https://groups.google.com/a/continuum.io/forum/#!topic/numba-users/DcWYd7IR5HE

**Comment by chcgnzls** [comment]
I don't think it is a broken installation or driver issue: I was able to compile and run some C programs without issue.

**Comment by abergeron** [comment]
The problem with CUDA is that it is comprised of many different parts and you can have some of them working and others not.  For example distro packages tend compute to various degrees since they don't test that functionality, but graphics usually work.  In your case it could be that in-process compilation doesn't work, but pre-compilation does since those two don't use the same codepath.

I'm not sure what the problem is exactly but it seems to be linked with the installation or driver version. I have never been able to reproduce it, so there is nothing I can do to fix it.

**Comment by chcgnzls** [comment]
It's just a little challenging for me to believe that among hundreds of other people using CUDA on this cluster, I'm the first person to run into part of a broken installation or driver error. Is there any kind of log file that would give me more information?

**Comment by abergeron** [comment]
I'm pretty sure that nobody does dynamic linking (except for numba apparently).  You can try to run some numba code to see if you have the same problem with that.

As far as log files, there aren't any.  All the "log" output is sent to stderr.  You can try to run a debug build of libgpuarray and that *could* give you more information, but I'm not sure.

**Comment by chcgnzls** [comment]
Thanks, I'll try those suggestions.

Minor developments: 
I have tried using other versions of cuda and have been getting the same error message. I also used the same installation method to put cuda on my personal machine that was used for the cluster (a *.run file directly from nvidia) and I **do not** get the same error using the *same cuda libraries and nvidia drivers (387.26)* as the cluster. However, we have different hardware: GeForce GTX 860M while the cluster has a Tesla K20Xm.
